From b504e6a79dd09631a67bb967de492755dd055b64 Mon Sep 17 00:00:00 2001
From: Ravindra Lokhande <rlokhande@nvidia.com>
Date: Thu, 9 Aug 2012 23:14:40 +0530
Subject: [PATCH] usbaudio: Get card number from substring

Use tinyalsa function to get card number from substring. USB audio
device card number can change as it can be plug-in/plug-out anytime.

Bug 1009921
Bug 1026047

Change-Id: Ic9e444e6970dc73e0c07605afff2d6ab7375a4b5

usbaudio: search card by audiosource substring

USB accessory uses separate driver which registers card with name
audiosource. First search card with USB substring if it fails then
search for audiosource substring.

Bug 1026047
Bug 1009921

Change-Id: I1ba4e3b609f4fb6c468823aca765d11b9deea0bc

audio: set framesize to 1 byte for MP3 & AAC format

Bug 967185

Change-Id: I9c6de96f3e1ac7d376abe8ca4789bf9b61b8cc10
---
 include/hardware/audio.h    |    3 +++
 modules/usbaudio/audio_hw.c |   16 ++++++++++++++--
 2 files changed, 17 insertions(+), 2 deletions(-)

diff --git a/include/hardware/audio.h b/include/hardware/audio.h
index 3a0962e..40cd231 100644
--- a/include/hardware/audio.h
+++ b/include/hardware/audio.h
@@ -301,6 +301,9 @@ static inline size_t audio_stream_frame_size(const struct audio_stream *s)
     case AUDIO_FORMAT_PCM_16_BIT:
         chan_samp_sz = sizeof(int16_t);
         break;
+    case AUDIO_FORMAT_MP3:
+    case AUDIO_FORMAT_AAC:
+        return sizeof(int8_t);
     case AUDIO_FORMAT_PCM_8_BIT:
     default:
         chan_samp_sz = sizeof(int8_t);
diff --git a/modules/usbaudio/audio_hw.c b/modules/usbaudio/audio_hw.c
index f33c343..9f0af93 100644
--- a/modules/usbaudio/audio_hw.c
+++ b/modules/usbaudio/audio_hw.c
@@ -252,12 +252,24 @@ static int adev_open_output_stream(struct audio_hw_device *dev,
 {
     struct audio_device *adev = (struct audio_device *)dev;
     struct stream_out *out;
+    int card_num = -1;
     int ret;
 
     out = (struct stream_out *)calloc(1, sizeof(struct stream_out));
     if (!out)
         return -ENOMEM;
 
+    card_num = snd_card_get_card_id_from_sub_string("USB");
+    if (card_num < 0) {
+        card_num = snd_card_get_card_id_from_sub_string("audiosource");
+        if (card_num < 0) {
+            ALOGV("Invalid card ID");
+            ret = -EINVAL;
+            goto err_open;
+        }
+    }
+    ALOGV("USB Card number %d \n", card_num);
+
     out->stream.common.get_sample_rate = out_get_sample_rate;
     out->stream.common.set_sample_rate = out_set_sample_rate;
     out->stream.common.get_buffer_size = out_get_buffer_size;
@@ -284,8 +296,8 @@ static int adev_open_output_stream(struct audio_hw_device *dev,
 
     out->standby = true;
 
-    adev->card = -1;
-    adev->device = -1;
+    adev->card = card_num;
+    adev->device = 0;
 
     *stream_out = &out->stream;
     return 0;
-- 
1.7.1

